src/main/java/com/google/jenkins/plugins/computeengine/ComputeEngineCloud.java

```diff
@@ -136,45 +134,67 @@ public String getDisplayName() {
   }
 
   protected Object readResolve() {
-    try {
-      ClientFactory clientFactory =
-          new ClientFactory(Jenkins.get(), new ArrayList<DomainRequirement>(), credentialsId);
-      this.client = clientFactory.compute();
-    } catch (IOException e) {
-      this.client = null;
-      // TODO: log
-    }
-
-    for (InstanceConfiguration c : configurations) {
-      c.cloud = this;
-      // Apply a label that associates an instance configuration with
-      // this cloud provider
-      c.appendLabel(CLOUD_ID_LABEL_KEY, getInstanceId());
-
-      // Apply a label that identifies the name of this instance configuration
-      c.appendLabel(CONFIG_LABEL_KEY, c.namePrefix);
+    if (configurations != null) {
+      for (InstanceConfiguration configuration : configurations) {
+        configuration.cloud = this;
+        configuration.readResolve();
+        // Apply a label that associates an instance configuration with
+        // this cloud provider
+        configuration.appendLabel(CLOUD_ID_LABEL_KEY, getInstanceId());
+
+        // Apply a label that identifies the name of this instance configuration
+        configuration.appendLabel(CONFIG_LABEL_KEY, configuration.namePrefix);
+      }
     }
     return this;
   }
 
+  /**
+   * Sets unique ID of that cloud instance.
+   *
+   * <p>This ID allows us to find machines from our cloud in GCP. <b>This value should not change
+   * between config reload, or nodes may be lost in GCP side</b>
+   */
+  @DataBoundSetter
+  public void setInstanceId(String instanceId) {
+    this.instanceId = instanceId;
+  }
+
   /**
    * Returns unique ID of that cloud instance.
    *
-   * <p>This ID allows us to find machines from our cloud in GCP. Current implementation is bit
-   * naive and generate ID based on name hashCode.
+   * <p>This ID allows us to find machines from our cloud in GCP.
    *
    * @return instance unique ID
    */
   public String getInstanceId() {
     return instanceId;
   }
 
+  private ComputeClient createClient() {
```

Comment:
This method only has a single call site. Please in-line it at the call site.

Reply:
I personally think that way more readable/corect..
That method that calls it have locking and it should be as clear as possible..

This way we have clear responsibility of methods..